proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -368,174 +368,95 @@ lemma invoke_sched_context_corres:
-       setSchedContext scPtr (scRefills_update (\<lambda>_. replaceAt 0 (scRefills sc) (Refill curTime budget))
-                              (scRefillMax_update (\<lambda>_. maxRefills)
-                               (scRefillCount_update (\<lambda>_. 1)
-                                (scRefillHead_update (\<lambda>_. 0) sc))));
-       maybeAddEmptyTail scPtr
-    od)"
-  apply (clarsimp simp: refillNew_def)
-  apply (rule monadic_rewrite_bind_tail[rotated])
-   apply (fastforce simp: getCurTime_def intro: gets_inv)
-  apply (rule monadic_rewrite_bind_tail[rotated, OF get_sc_inv'])
-  apply (rule monadic_rewrite_bind_tail[rotated, OF set_sc'.typ_at_lifts'(6)])
-  apply (rule monadic_rewrite_bind_tail[rotated, OF get_sc_inv'])
-  apply (rule monadic_rewrite_bind_tail[rotated, OF set_sc'.typ_at_lifts'(6)])
-  apply (simp flip: bind_assoc)
-  apply (rule monadic_rewrite_bind_head)
-  apply (simp add: bind_assoc)
-  apply (subst bind_assoc_group4)
-  apply (rule monadic_rewrite_rewrite_head)
-   apply (rule monadic_rewrite_imp)
-    apply (rule monadic_rewrite_sym)
-    apply (rule getSchedContext_setSchedContext_decompose_decompose_ext2[unfolded K_bind_def])
-   apply (clarsimp simp: objBits_simps)
-  apply (simp only: bind_assoc)
-  apply (subst bind_assoc_group4)
-  apply (rule monadic_rewrite_rewrite_head)
-   apply (rule monadic_rewrite_imp)
-    apply (rule monadic_rewrite_sym)
-    apply (rule getSchedContext_setSchedContext_decompose_decompose_ext2[unfolded K_bind_def])
-   apply (clarsimp simp: objBits_simps)
-  apply (simp only: bind_assoc)
-  apply (clarsimp simp: setRefillHd_def updateRefillHd_def bind_assoc)
-  apply (rule monadic_rewrite_imp)
-   apply (rule monadic_rewrite_trans_dup)
-    apply (rule monadic_rewrite_sym)
-    apply (rule getSchedContext_setSchedContext_decompose[unfolded K_bind_def])
-   apply (rule monadic_rewrite_refl3)
-   apply clarsimp
-  apply (fastforce simp: objBits_simps length_replaceAt)
-  done
-
 lemma getCurTime_sp:
   "\<lbrace>P\<rbrace> getCurTime \<lbrace>\<lambda>rv. P and (\<lambda>s. rv = ksCurTime s)\<rbrace>"
   by (wpsimp simp: getCurTime_def)
 
 lemma isRoundRobin_corres:
-  "corres (=) (pspace_aligned and pspace_distinct and sc_at sc_ptr) \<top>
+  "corres (=) (sc_at sc_ptr) (sc_at' sc_ptr)
               (is_round_robin sc_ptr) (isRoundRobin sc_ptr)"
-  apply (rule corres_cross[where Q' = "sc_at' sc_ptr", OF sc_at'_cross_rel], fastforce)
   apply (clarsimp simp: is_round_robin_def isRoundRobin_def)
   apply (corressimp corres: get_sc_corres
                       simp: sc_relation_def)
   done
 
+lemma valid_obj'_scPeriod_update[simp]:
+  "valid_obj' (KOSchedContext (scPeriod_update (\<lambda>_. period) sc')) = valid_obj' (KOSchedContext sc')"
+  by (fastforce simp: valid_obj'_def valid_sched_context'_def valid_sched_context_size'_def objBits_simps)
+
+(* it would be good to rewrite all getting wp rules in this form *)
+lemma getSchedContext_wp':
```

Comment:
This is now a duplicate, I should remove (now generated from the `simple_ko’` locale in KHeap_R).

Reply:
Are you planning to remove it from this PR, or in a future commit?

Reply:
removed